### ISSUE TITLE
overloaded function passed into function with ParamSpec error

### ISSUE THREAD

**Issue by NelschW** [microsoft/pyright#5631]
**Describe the bug**
Pyright reports an error for the example, but it should not.
Also the type hint for f2 in the first code segment missing/wrong.

**Code or Screenshots**

```python
from typing import Callable, TypeVar,ParamSpec, overload

P=ParamSpec("P")
T=TypeVar("T")


@overload
def f(x:str)->str:...
@overload
def f(x:int)->int:...

def f(x:str|int):
    return x

def get_f2(f:Callable[P,T]):
    def f2(*args:P.args,**kwargs:P.kwargs) -> list[T]:
        return [f(*args,**kwargs)]
    return f2

f2=get_f2(f)

f2(1) # error display but should be valid
```

without @overload everything is fine

```python
from typing import Callable, TypeVar,ParamSpec, overload

P=ParamSpec("P")
T=TypeVar("T")




def f(x:str|int):
    return x

def get_f2(f:Callable[P,T]):
    def f2(*args:P.args,**kwargs:P.kwargs) -> list[T]:
        return [f(*args,**kwargs)]
    return f2

f2=get_f2(f)

f2(1) # works fine
```

Pylance: `v2023.7.43`

**Comment by erictraut** [comment]
If you add a return type annotation to `get_f2`, the error goes away.

```
def get_f2(f: Callable[P, T]) -> Callable[P, list[T]]: ...
```

I'll investigate why the inferred type in this case is not working the same as the declared type.

Capturing overloaded functions with a ParamSpec is complicated. PEP 612 (which introduced ParamSpec) doesn't explain how it should work. Mypy doesn't even attempt to support it. I've spent a bunch of time in pyright to make it work in most cases. It should work fine here.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.321, which I just published. It will also be included in a future release of pylance.
